hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java

```diff
@@ -3441,54 +3437,73 @@ public String formatBinaryLiteral(byte[] bytes) {
 
 	/**
 	 * Pluggable strategy for determining the Size to use for columns of
-	 * a given SQL type when no explicit Size has been given.
+	 * a given SQL type.
 	 *
-	 * Allows Dialects, integrators and users a chance to apply default
-	 * column size limits in certain situations based on the mapped
+	 * Allows Dialects, integrators and users a chance to apply
+	 * column size defaults and limits in certain situations based on the mapped
 	 * SQL and Java types.  E.g. when mapping a UUID to a VARCHAR column
 	 * we know the default Size should be `Size#length == 36`.
 	 */
-	public interface DefaultSizeStrategy {
+	public interface SizeStrategy {
 		/**
-		 * Resolve the default {@link Size} to use for columns of the given
+		 * Resolve the {@link Size} to use for columns of the given
 		 * {@link SqlTypeDescriptor SQL type} and {@link JavaTypeDescriptor Java type}.
 		 *
 		 * @return a non-null {@link Size}
 		 */
-		Size resolveDefaultSize(SqlTypeDescriptor sqlType, JavaTypeDescriptor javaType);
+		Size resolveSize(
+				SqlTypeDescriptor sqlType,
+				JavaTypeDescriptor javaType,
+				Integer precision,
+				Integer scale, Long length);
 	}
 
-	public class DefaultSizeStrategyImpl implements DefaultSizeStrategy {
+	public class SizeStrategyImpl implements SizeStrategy {
 		@Override
-		public Size resolveDefaultSize(SqlTypeDescriptor sqlType, JavaTypeDescriptor javaType) {
+		public Size resolveSize(
+				SqlTypeDescriptor sqlType,
+				JavaTypeDescriptor javaType,
+				Integer precision,
+				Integer scale,
+				Long length) {
 			final Size size = new Size();
 			int jdbcTypeCode = sqlType.getJdbcTypeCode();
 
 			switch (jdbcTypeCode) {
 				case Types.BIT:
+					// BIT has an implicit length of 1
+					// If a database like e.g. MySQL supports values greater than 1, it can override this
+					length = null;
+					precision = null;
+					scale = null;
+					break;
 				case Types.CHAR:
 				case Types.NCHAR:
 				case Types.BINARY:
 				case Types.VARCHAR:
 				case Types.NVARCHAR:
 				case Types.VARBINARY:
-					size.setLength( javaType.getDefaultSqlLength(Dialect.this) );
+					size.setLength( javaType.getDefaultSqlLength( Dialect.this ) );
 					break;
 				case Types.LONGVARCHAR:
 				case Types.LONGNVARCHAR:
 				case Types.LONGVARBINARY:
 					size.setLength( javaType.getLongSqlLength() );
 					break;
 				case Types.FLOAT:
+					if ( precision != null && precision < 1 ) {
+						// Use the default precision
+						precision = null;
+					}
```

Comment:
When the JPA `@Column` annotation is used, which has a default precision of 0, we receive a precision of 0 here, but e.g. a `float(0)` is illegal.

Reply:
OK but this seems like the wrong layer to handle interpretation of `@Column`. Surely it should be the job of the `Binder` to interpret 0 as `null` when reading the annotation.

Reply:
For precision, you might be right that we should interpret 0 as null to get the default precision, but for `scale` it's more complicated as you can read in the discussion: https://hibernate.zulipchat.com/login/#narrow/stream/132094-hibernate-orm-dev/topic/Minimum.20sizes.20for.20types

My main argument for interpreting the precision here is to have a single place to interpret all size related values.

Reply:
Honestly `scale` I'm simply not sure about. I don't think there's a clear intent in the JPA spec, and I don't recall us ever thinking terribly hard about this. I think we just accepted that there were no nulls in annotations so defaulting to `0` was the best we could possibly do.

*Probably* `scale=0` should be taken literally.

But that's not the case with precision, since `precision=0` simply makes no sense.

Unfortunate, but anyway with the `getDefaultSqlPrecision()` and `getDefaultSqlScale()` stuff I implemented, explicitly specifying precision and scale less-often needed.

Reply:
By the way, @beikov, have you checked that none of these changes break my work on defaulting the right precision/scale/length for a type?

Reply:
Break "how" and which work are you referring to? If there are tests for that, this is covered.

I'll change the binder code to handle precision 0 the same as `null` then.

Reply:
Well I mean all the work I did to make sure that we generate DDL with a good precision/scale by default for each Hibernate type. (In Hibernate 5 the precision/scale often defaulted to rubbish.)

I don't think there's tests, because I didn't have a good way to test the types in generated DDL.

Reply:
Since I reverted the changes regarding the minimum size check, this is equivalent to the previous behavior now.

Reply:
OK, 👍 thanks for confirming